File: astar.py
# https://github.com/zhm-real/MotionPlanning

import heapq
import math
import numpy as np
from dataclasses import dataclass
import rerun as rr


@dataclass
class HolonomicNode:
    x: int  # x position of node
    y: int  # y position of node
    cost: float  # g cost of node
    pind: int  # parent index of node


@dataclass
class Param:
    minx: int
    miny: int
    maxx: int
    maxy: int
    xw: int
    yw: int
    reso: float  # resolution of grid world
    motion: list[tuple[int, int]]  # motion set


def astar_planning(
    sx: float,
    sy: float,
    gx: float,
    gy: float,
    ox: list[float],
    oy: list[float],
    reso: float,
    robo_radius: float,
) -> tuple[list[float], list[float]] | None:
    """
    return path of A*.
    :param sx: starting node x [m]
    :param sy: starting node y [m]
    :param gx: goal node x [m]
    :param gy: goal node y [m]
    :param ox: obstacles x positions [m]
    :param oy: obstacles y positions [m]
    :param reso: xy grid resolution
    :param rr: robot radius
    :return: path
    """

    n_start = HolonomicNode(int(np.round(sx / reso)), int(np.round(sy / reso)), 0.0, -1)
    n_goal = HolonomicNode(int(np.round(gx / reso)), int(np.round(gy / reso)), 0.0, -1)
    ox = [x / reso for x in ox]  # in grid, float
    oy = [y / reso for y in oy]  # in grid, float

    param, obsmap = calc_parameters(ox, oy, robo_radius, reso)
    goal_ind = calc_index(n_goal, param)

    open_set, closed_set = dict(), dict()
    open_set[calc_index(n_start, param)] = n_start

    q_priority = []
    heapq.heappush(q_priority, (fvalue(n_start, n_goal), calc_index(n_start, param)))

    while len(open_set) != 0:
        _, ind = heapq.heappop(q_priority)
        n_curr = open_set[ind]
        closed_set[ind] = n_curr
        open_set.pop(ind)
        if goal_ind == ind:
            exps = []
            for n in closed_set.values():
                exps.append([n.x, n.y])
            rr.log("explored", rr.Points2D(np.array(exps)))
            pathx, pathy = extract_path(closed_set, n_start, n_goal, param)
            return pathx, pathy

        for i in range(len(param.motion)):
            node = HolonomicNode(
                n_curr.x + param.motion[i][0],
                n_curr.y + param.motion[i][1],
                n_curr.cost + u_cost(param.motion[i]),
                ind,
            )

            if not check_node(node, param, obsmap):
                continue

            n_ind = calc_index(node, param)

            if n_ind not in closed_set:
                if n_ind in open_set:
                    if open_set[n_ind].cost > node.cost:
                        open_set[n_ind].cost = node.cost
                        open_set[n_ind].pind = ind
                else:
                    open_set[n_ind] = node
                    heapq.heappush(
                        q_priority, (fvalue(node, n_goal), calc_index(node, param))
                    )

    exps = []
    for n in closed_set.values():
        exps.append([n.x, n.y])
    rr.log("explored", rr.Points2D(np.array(exps)))
    return None


def calc_holonomic_heuristic_with_obstacle(
    node, ox: list[float], oy: list[float], reso: float, rr: float
) -> list[list[float]]:
    n_goal = HolonomicNode(
        int(np.round(node.x[-1] / reso)), int(np.round(node.y[-1] / reso)), 0.0, -1
    )

    ox = [x / reso for x in ox]
    oy = [y / reso for y in oy]

    P, obsmap = calc_parameters(ox, oy, rr, reso)

    open_set, closed_set = dict(), dict()
    open_set[calc_index(n_goal, P)] = n_goal

    q_priority = []
    heapq.heappush(q_priority, (n_goal.cost, calc_index(n_goal, P)))

    while open_set:
        _, ind = heapq.heappop(q_priority)
        n_curr = open_set[ind]
        closed_set[ind] = n_curr
        open_set.pop(ind)

        for mot in P.motion:
            node = HolonomicNode(
                n_curr.x + mot[0],
                n_curr.y + mot[1],
                n_curr.cost + u_cost(mot),
                ind,
            )

            if not check_node(node, P, obsmap):
                continue

            n_ind = calc_index(node, P)
            if n_ind not in closed_set:
                if n_ind in open_set:
                    if open_set[n_ind].cost > node.cost:
                        open_set[n_ind].cost = node.cost
                        open_set[n_ind].pind = ind
                else:
                    open_set[n_ind] = node
                    heapq.heappush(q_priority, (node.cost, calc_index(node, P)))

    hmap = [[np.inf for _ in range(P.yw)] for _ in range(P.xw)]

    for n in closed_set.values():
        hmap[n.x - P.minx][n.y - P.miny] = n.cost

    return hmap


def check_node(node: HolonomicNode, P: Param, obsmap: list[list[bool]]) -> bool:
    if node.x <= P.minx or node.x >= P.maxx or node.y <= P.miny or node.y >= P.maxy:
        return False

    if obsmap[node.x - P.minx][node.y - P.miny]:
        return False

    return True


def u_cost(u: tuple[int, int]) -> float:
    return math.hypot(u[0], u[1])


def fvalue(node: HolonomicNode, n_goal: HolonomicNode) -> float:
    return node.cost + h(node, n_goal)


def h(node: HolonomicNode, n_goal: HolonomicNode) -> float:
    return math.hypot(node.x - n_goal.x, node.y - n_goal.y)


def calc_index(node: HolonomicNode, P: Param) -> int:
    return (node.y - P.miny) * P.xw + (node.x - P.minx)


def calc_parameters(
    ox: list[float], oy: list[float], rr: float, reso: float
) -> tuple[Param, list[list[bool]]]:
    minx, miny = int(np.round(min(ox))), int(np.round(min(oy)))
    maxx, maxy = int(np.round(max(ox))), int(np.round(max(oy)))
    xw, yw = maxx - minx, maxy - miny

    motion = get_motion()
    P = Param(minx, miny, maxx, maxy, xw, yw, reso, motion)
    obsmap = calc_obsmap(ox, oy, rr, P)

    return P, obsmap


def calc_obsmap(
    ox: list[float], oy: list[float], rr: float, P: Param
) -> list[list[bool]]:
    obsmap = [[False for _ in range(P.yw)] for _ in range(P.xw)]

    for x in range(P.xw):
        xx = x + P.minx
        for y in range(P.yw):
            yy = y + P.miny
            for oxx, oyy in zip(ox, oy):
                if math.hypot(oxx - xx, oyy - yy) <= rr / P.reso:
                    obsmap[x][y] = True
                    break

    return obsmap


def extract_path(
    closed_set: dict, n_start: dict, n_goal: HolonomicNode, P: Param
) -> tuple[list[float], list[float]]:
    pathx, pathy = [n_goal.x], [n_goal.y]
    n_ind = calc_index(n_goal, P)

    while True:
        node = closed_set[n_ind]
        pathx.append(node.x)
        pathy.append(node.y)
        n_ind = node.pind

        if node == n_start:
            break

    pathx = [x * P.reso for x in reversed(pathx)]
    pathy = [y * P.reso for y in reversed(pathy)]

    return pathx, pathy


def get_motion():
    motion = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]

    return motion


def get_env() -> tuple[list[int], list[int]]:
    ox, oy = [], []

    for i in range(60):
        ox.append(i)
        oy.append(0.0)
    for i in range(60):
        ox.append(60.0)
        oy.append(i)
    for i in range(61):
        ox.append(i)
        oy.append(60.0)
    for i in range(61):
        ox.append(0.0)
        oy.append(i)
    for i in range(40):
        ox.append(20.0)
        oy.append(i)
    for i in range(40):
        ox.append(40.0)
        oy.append(60.0 - i)

    return ox, oy


def main():
    rr.init("astar")
    rr.connect_tcp("172.28.64.1:9876")

    sx = 10.0  # [m]
    sy = 10.0  # [m]
    gx = 50.0  # [m]
    gy = 50.0  # [m]

    robot_radius = 2.0
    grid_resolution = 1.0
    ox, oy = get_env()
    rr.log("obstacles", rr.Points2D(np.array((ox, oy)).T))
    rr.log("start", rr.Points2D((sx, sy)))
    rr.log("goal", rr.Points2D((gx, gy)))

    res = astar_planning(sx, sy, gx, gy, ox, oy, grid_resolution, robot_radius)
    if res is None:
        print("Searching failed!")
        return
    pathx, pathy = res
    rr.log("path", rr.LineStrips2D(np.array((pathx, pathy)).T))


if __name__ == "__main__":
    main()

File: src/astar.rs
use crate::data_struct::Node;
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};
use std::f64;

use ordered_float::NotNan;
use rerun::external::glam::Vec2;

#[derive(Debug, Clone, Copy, PartialEq)]
struct HolonomicNode {
    x: usize,    // x position of node
    y: usize,    // y position of node
    cost: f64,   // g cost of node
    pind: isize, // parent index of node
}

#[derive(Debug, Clone)]
struct Param {
    minx: usize,
    miny: usize,
    maxx: usize,
    maxy: usize,
    xw: usize,
    yw: usize,
    reso: f64,                   // resolution of grid world
    motion: Vec<(isize, isize)>, // motion set
}

impl HolonomicNode {
    fn new(x: usize, y: usize, cost: f64, pind: isize) -> Self {
        HolonomicNode { x, y, cost, pind }
    }
}

impl Param {
    fn new(minx: usize, miny: usize, maxx: usize, maxy: usize, reso: f64) -> Self {
        let xw = maxx - minx;
        let yw = maxy - miny;
        let motion = get_motion();
        Param {
            minx,
            miny,
            maxx,
            maxy,
            xw,
            yw,
            reso,
            motion,
        }
    }
}

pub fn astar_planning(
    sx: f64,
    sy: f64,
    gx: f64,
    gy: f64,
    ox: &[f64],
    oy: &[f64],
    reso: f64,
    rr: f64,
    rec: &rerun::RecordingStream,
) -> Option<(Vec<f64>, Vec<f64>)> {
    let n_start = HolonomicNode::new(
        (sx / reso).round() as usize,
        (sy / reso).round() as usize,
        0.0,
        -1,
    );
    let n_goal = HolonomicNode::new(
        (gx / reso).round() as usize,
        (gy / reso).round() as usize,
        0.0,
        -1,
    );

    let ox: Vec<f64> = ox.iter().map(|&x| x / reso).collect();
    let oy: Vec<f64> = oy.iter().map(|&y| y / reso).collect();

    let (param, obsmap) = calc_parameters(&ox, &oy, rr, reso);
    let goal_ind = calc_index(&n_goal, &param);
    let mut open_set: HashMap<usize, HolonomicNode> = HashMap::new();
    let mut closed_set: HashMap<usize, HolonomicNode> = HashMap::new();
    open_set.insert(calc_index(&n_start, &param), n_start);

    let mut q_priority = BinaryHeap::new();
    q_priority.push((
        Reverse(NotNan::new(fvalue(n_start, n_goal)).unwrap()),
        calc_index(&n_start, &param),
    ));
    let mut explored: Vec<Vec2> = Vec::new();

    while !open_set.is_empty() {
        let (_, ind) = q_priority.pop().unwrap();
        let n_curr = open_set.remove(&ind).unwrap();
        closed_set.insert(ind, n_curr);
        explored.push(Vec2::new(n_curr.x as f32, n_curr.y as f32));
        let _ = rec.log("explored", &rerun::Points2D::new(explored.clone()));
        if goal_ind == ind {
            return Some(extract_path(&closed_set, &n_start, &n_goal, &param));
        }

        for motion in &param.motion {
            let node = HolonomicNode::new(
                (n_curr.x as isize + motion.0) as usize,
                (n_curr.y as isize + motion.1) as usize,
                n_curr.cost + u_cost(*motion),
                ind as isize,
            );

            if !check_node(&node, &param, &obsmap) {
                continue;
            }

            let n_ind = calc_index(&node, &param);
            if !closed_set.contains_key(&n_ind) {
                if let Some(open_node) = open_set.get_mut(&n_ind) {
                    if open_node.cost > node.cost {
                        open_node.cost = node.cost;
                        open_node.pind = ind as isize;
                    }
                } else {
                    open_set.insert(n_ind, node);
                    q_priority.push((Reverse(NotNan::new(fvalue(node, n_goal)).unwrap()), n_ind));
                }
            }
        }
    }
    None
}

pub fn calc_holonomic_heuristic_with_obstacle(
    node: &Node,
    ox: &Vec<f64>,
    oy: &Vec<f64>,
    reso: f64,
    rr: f64,
) -> Vec<Vec<f64>> {
    let n_goal = HolonomicNode::new(
        (node.x.last().unwrap() / reso).round() as usize,
        (node.y.last().unwrap() / reso).round() as usize,
        0.0,
        -1,
    );

    let oxf: Vec<f64> = ox.iter().map(|&x| x / reso).collect();
    let oyf: Vec<f64> = oy.iter().map(|&x| x / reso).collect();

    let (param, obsmap) = calc_parameters(&oxf, &oyf, rr, reso);

    let mut open_set = HashMap::new();
    let mut closed_set = HashMap::new();
    open_set.insert(calc_index(&n_goal, &param), n_goal);

    let mut q_priority = BinaryHeap::new();
    q_priority.push((
        Reverse(NotNan::new(n_goal.cost).unwrap()),
        calc_index(&n_goal, &param),
    ));

    while !open_set.is_empty() {
        let (_, ind) = q_priority.pop().unwrap();
        let n_curr = open_set.remove(&ind).unwrap();
        closed_set.insert(ind, n_curr);
        for cmd in param.motion.clone() {
            let node = HolonomicNode::new(
                (n_curr.x as isize + cmd.0) as usize,
                (n_curr.y as isize + cmd.1) as usize,
                n_curr.cost + u_cost(cmd),
                ind as isize,
            );

            if !check_node(&node, &param, &obsmap) {
                continue;
            }
            let n_ind = calc_index(&node, &param);
            if !closed_set.contains_key(&n_ind) {
                if let Some(open_node) = open_set.get_mut(&n_ind) {
                    if open_node.cost > node.cost {
                        open_node.cost = node.cost;
                        open_node.pind = ind as isize;
                    }
                } else {
                    open_set.insert(n_ind, node);
                    q_priority.push((Reverse(NotNan::new(node.cost).unwrap()), n_ind));
                }
            }
        }
    }

    let mut hmap = vec![vec![f64::INFINITY; param.yw]; param.xw];
    closed_set.into_values().for_each(|n| {
        hmap[n.x - param.minx][n.y - param.miny] = n.cost;
    });
    hmap
}

fn check_node(node: &HolonomicNode, param: &Param, obsmap: &[Vec<bool>]) -> bool {
    if node.x <= param.minx || node.x >= param.maxx || node.y <= param.miny || node.y >= param.maxy
    {
        return false;
    }

    if obsmap[node.x - param.minx][node.y - param.miny] {
        return false;
    }

    true
}

fn u_cost(u: (isize, isize)) -> f64 {
    ((u.0 * u.0 + u.1 * u.1) as f64).sqrt()
}

fn fvalue(node: HolonomicNode, n_goal: HolonomicNode) -> f64 {
    node.cost + h(node, n_goal)
}

fn h(node: HolonomicNode, n_goal: HolonomicNode) -> f64 {
    (((node.x as isize - n_goal.x as isize).pow(2) + (node.y as isize - n_goal.y as isize).pow(2))
        as f64)
        .sqrt()
}

fn calc_index(node: &HolonomicNode, param: &Param) -> usize {
    (node.y - param.miny) * param.xw + (node.x - param.minx)
}

fn calc_parameters(ox: &Vec<f64>, oy: &Vec<f64>, rr: f64, reso: f64) -> (Param, Vec<Vec<bool>>) {
    let minx = (ox.iter().cloned().fold(f64::INFINITY, f64::min)).round() as usize;
    let miny = (oy.iter().cloned().fold(f64::INFINITY, f64::min)).round() as usize;
    let maxx = (ox.iter().cloned().fold(f64::NEG_INFINITY, f64::max)).round() as usize;
    let maxy = (oy.iter().cloned().fold(f64::NEG_INFINITY, f64::max)).round() as usize;

    let param = Param::new(minx, miny, maxx, maxy, reso);
    let obsmap = calc_obsmap(ox, oy, rr, &param);

    (param, obsmap)
}

fn calc_obsmap(ox: &Vec<f64>, oy: &Vec<f64>, rr: f64, param: &Param) -> Vec<Vec<bool>> {
    let mut obsmap = vec![vec![false; param.yw]; param.xw];

    for x in 0..param.xw {
        let xx = x + param.minx;
        for y in 0..param.yw {
            let yy = y + param.miny;
            for (&oxx, &oyy) in ox.iter().zip(oy.iter()) {
                if ((oxx - xx as f64).hypot(oyy - yy as f64)) <= rr / param.reso {
                    obsmap[x][y] = true;
                    break;
                }
            }
        }
    }

    obsmap
}

fn extract_path(
    closed_set: &HashMap<usize, HolonomicNode>,
    n_start: &HolonomicNode,
    n_goal: &HolonomicNode,
    param: &Param,
) -> (Vec<f64>, Vec<f64>) {
    let mut pathx = vec![n_goal.x as f64];
    let mut pathy = vec![n_goal.y as f64];
    let mut n_ind = calc_index(n_goal, param);

    while let Some(node) = closed_set.get(&n_ind) {
        pathx.push(node.x as f64);
        pathy.push(node.y as f64);
        n_ind = node.pind as usize;

        if node == n_start {
            break;
        }
    }

    pathx.reverse();
    pathy.reverse();

    for x in &mut pathx {
        *x *= param.reso;
    }

    for y in &mut pathy {
        *y *= param.reso;
    }

    (pathx, pathy)
}

fn get_motion() -> Vec<(isize, isize)> {
    vec![
        (-1, 0),
        (-1, 1),
        (0, 1),
        (1, 1),
        (1, 0),
        (1, -1),
        (0, -1),
        (-1, -1),
    ]
}

pub fn get_env() -> (Vec<f64>, Vec<f64>) {
    let mut ox = Vec::new();
    let mut oy = Vec::new();

    for i in 0..60 {
        ox.push(i as f64);
        oy.push(0.0);
    }
    for i in 0..60 {
        ox.push(60.0);
        oy.push(i as f64);
    }
    for i in 0..61 {
        ox.push(i as f64);
        oy.push(60.0);
    }
    for i in 0..61 {
        ox.push(0.0);
        oy.push(i as f64);
    }
    for i in 0..40 {
        ox.push(20.0);
        oy.push(i as f64);
    }
    for i in 0..40 {
        ox.push(40.0);
        oy.push(60.0 - i as f64);
    }

    (ox, oy)
}
